rust/ethers-prometheus/src/lib.rs

```diff
@@ -0,0 +1,518 @@
+
+        if let Some(m) = &self.metrics.contract_call_duration_seconds {
+            let data = self.data.read().await;
+            let chain_name = metrics_chain_name(tx.chain_id().map(|id| id.as_u64()));
+            let (contract_addr, contract_name) = tx
+                .to()
+                .and_then(|addr| match addr {
+                    NameOrAddress::Name(n) => Some((n.clone(), n.clone())),
+                    NameOrAddress::Address(a) => data
+                        .contracts
+                        .get(a)
+                        .and_then(|c| c.name.clone())
+                        .map(|n| (a.to_string(), n)),
+                })
+                .unwrap_or_else(|| ("".into(), "unknown".into()));
+
+            m.with(&hashmap! {
+                "chain" => chain_name.as_str(),
+                "contract_name" => contract_name.as_str(),
+                "contract_address" => contract_addr.as_str()
+            })
+            .observe((Instant::now() - start).as_secs_f64())
+        }
+
+        Ok(result?)
+    }
+}
+
+impl<M> PrometheusMiddleware<M> {
+    /// Create a new prometheus middleware instance.
+    /// - `inner`: The wrapped middleware.
+    /// - `metrics`: Metrics objects we will report to.
+    /// - `tokens`: Tokens to watch the balances of.
+    /// - `wallets`: Wallets to watch the balances of.
+    pub fn new(inner: M, metrics: ProviderMetrics, conf: PrometheusMiddlewareConf) -> Self {
+        Self {
+            inner: Arc::new(inner),
+            metrics,
+            data: Arc::new(RwLock::new(conf)),
+        }
+    }
+
+    /// Start tracking metrics for a new token.
+    pub async fn track_new_token(&self, addr: Address, info: TokenInfo) {
+        self.track_new_tokens([(addr, info)]).await;
+    }
+
+    /// Start tacking metrics for new tokens.
+    pub async fn track_new_tokens(&self, iter: impl IntoIterator<Item = (Address, TokenInfo)>) {
+        let mut data = self.data.write().await;
+        for (addr, info) in iter {
+            data.tokens.insert(addr, info);
+        }
+    }
+
+    /// Start tracking metrics for a new wallet.
+    pub async fn track_new_wallet(&self, addr: Address, info: WalletInfo) {
+        self.track_new_wallets([(addr, info)]).await;
+    }
+
+    /// Start tracking metrics for new wallets.
+    pub async fn track_new_wallets(&self, iter: impl IntoIterator<Item = (Address, WalletInfo)>) {
+        let mut data = self.data.write().await;
+        for (addr, info) in iter {
+            data.wallets.insert(addr, info);
+        }
+    }
+}
+
+impl<M: Middleware> PrometheusMiddleware<M> {
+    /// Start the update cycle using tokio. This must be called if you want
+    /// some metrics to be updated automatically. Alternatively you could call update yourself.
+    pub fn start_updating_on_interval(
+        self: &Arc<Self>,
+        period: Duration,
+    ) -> impl Future<Output = ()> + Send {
+        let zelf = Arc::downgrade(self);
```

Comment:
Ah yeah, so this is fun. Basically we are creating a long-running future that does a thing on an interval. We take `self: &Arc<self>` so that we can create a weak pointer, and then stop running when the weak reference ceases to resolve indicating all instances of the Arc were dropped.

We return a Future from a non-async fn because we want have different lifetimes than the arguments (namely self) and we achieve this by making an owned weak reference that we can move into the future. If it were an async fn we would have to have `self: Weak<Self>` which is not currently supported but Arc<Self> is (see custom receivers).

Basically this allows the caller to spin up this thread and then exit because the lifetime is not dependent on the caller.